_posts/2023-01-14-trivially-relocatable.md

```diff
@@ -0,0 +1,656 @@
+
+Since this behaviour is all implementation defined, we have unfortunate differences between
+compilers however, which can make testing for behaviour tricky. As in the
+[base class example](#base-classes), MSVC 19
+[does not](https://godbolt.org/z/TxoP4h9jY) make use of the tail padding in `S`
+and the size of `T` will be 12 as a result. This [also holds](https://godbolt.org/z/x6szhYPxo) when
+using the compiler-specific [`[[msvc::no_unique_address]]` attribute](
+https://devblogs.microsoft.com/cppblog/msvc-cpp20-and-the-std-cpp20-switch/) instead.
+
+##### Sorry but memcpy() with sizeof(T) is dangerous
+
+The outcome of the above is that `memcpy(dest, src, sizeof(T))` is dangerous in generic code, and
+even more so in C++20. The more correct thing to do is
+`memcpy(dest, src, sus::mem::data_size_of<T>())`.
+
+#### The Limits of Data size in a Library
+
+For some types, we can not determine a data size. In particular, a union's data size is dynamic
+depending on its active member. If we could enumerate the members of a union, we could use the
+maximum data size of all its members, but that is beyond the scope of what a library can achieve
+unfortunately.
+
+For that reason, the implementation of `sus::mem::data_size_of<T>()` on a union type
+[will return 0](https://github.com/chromium/subspace/blob/21a55214fdd968ba01118697721b708b83540521/subspace/mem/__private/data_size_finder.h#L61-L65).
+This means that a data size of 0 should be treated as unknown, and this explains why we require
+a non-zero data size in `sus::mem::relocate_by_memcpy`.
+
+#### Volatile and trivially relocatable
+
+Popping the stack, we were walking through our implementation of `sus::mem::relocate_by_memcpy`.
+The first condition checked is if the type is `volatile`. A `volatile` type can not be copied
+byte-by-byte with `memcpy()` without [introducing the chance of tearing](
+https://quuxplusone.github.io/blog/2018/07/13/trivially-copyable-corner-cases/). So these types
+are strictly excluded. This also helps with defining trivial relocatable for classes with volatile
+members as we will see.
+
+### Opting into trivially relocatable
+
+In Clang, a type may opt into being trivially relocatable with the class annotation
+`[[clang::trivial_abi]]`. However there are two important things missing from this attribute:
+
+1. The attribute only works on Clang. A standard library should work well across all compilers.
+1. The attribute can't take template parameters into account.
+
+While `[[clang::trivial_abi]]` works for `std::unique_ptr<T>`, it does so because the data member of
+the class is `T*` and a pointer is always trivially relocatable. A simple example where we can not
+use the attribute would be:
+
+```cpp
+template <class T>
+struct S {
+    T t;
+};
+```
+
+Here the struct `S` is trivially relocatable if `T` is. But marking the type with
+`[[clang::trivial_abi]]` would introduce Undefined Behaviour when `T` is not. And this is where the

```

Comment:
Maybe this is discussed more in one of the earlier context links; if not, it might be nice to make undefined behavior provide a context link to explain why this is

Reply:
Yeah, it's not elsewhere so let me talk a little bit about it, thanks. (spoiler: it means your `this` pointer can change during your object's lifetime)